glue/core/state.py

```diff
@@ -1263,3 +1275,170 @@ def apply_inplace_patches(rec):
+            continue
+        else:
+            meta_filtered[key] = value
+    result["meta"] = context.do(meta_filtered)
+
+    return result
+
+
+@loader(RegionData, version=1)
+def _load_regiondata(rec, context):
+    """
+    Custom load function for RegionData.
+    This is the same as the chain of logic in
+    _save_data_5 for Data, but result is an RegionData object
+    instead.
+    """
+
+    label = rec["label"]
+    result = RegionData(label=label)
+
+    # we manually rebuild pixel/world components, so
+    # we override this function. This is pretty ugly
+    result._create_pixel_and_world_components = lambda ndim: None
+
+    comps = [list(map(context.object, [cid, comp])) for cid, comp in rec["components"]]
+
+    for icomp, (cid, comp) in enumerate(comps):
+        if isinstance(comp, CoordinateComponent):
+            comp._data = result
+
+            # For backward compatibility, we need to check for cases where
+            # the component ID for the pixel components was not a PixelComponentID
+            # and upgrade it to one. This can be removed once we no longer
+            # support pre-v0.8 session files.
+            if not comp.world and not isinstance(cid, PixelComponentID):
+                cid = PixelComponentID(comp.axis, cid.label, parent=cid.parent)
+                comps[icomp] = (cid, comp)
+
+        result.add_component(comp, cid)
+
+    assert result._world_component_ids == []
+
+    coord = [c for c in comps if isinstance(c[1], CoordinateComponent)]
+    coord = [x[0] for x in sorted(coord, key=lambda x: x[1])]
+
+    if getattr(result, "coords") is not None:
+        assert len(coord) == result.ndim * 2
+        # Might black formatting break this?
+        result._world_component_ids = coord[: len(coord) // 2]
+        result._pixel_component_ids = coord[len(coord) // 2 :]  # noqa E203
+    else:
+        assert len(coord) == result.ndim
+        result._pixel_component_ids = coord
+
+    # We can now re-generate the coordinate links
+    result._set_up_coordinate_component_links(result.ndim)
+
+    for s in rec["subsets"]:
+        result.add_subset(context.object(s))
+
+    result.style = context.object(rec["style"])
+
+    if "primary_owner" in rec:
+        for cid in rec["primary_owner"]:
+            cid = context.object(cid)
+            cid.parent = result
+    yield result
+
+    def load_cid_tuple(cids):
+        return tuple(context.object(cid) for cid in cids)
+
+    result._key_joins = dict(
+        (context.object(k), (load_cid_tuple(v0), load_cid_tuple(v1)))
+        for k, v0, v1 in rec["_key_joins"]
+    )
+    if "uuid" in rec and rec["uuid"] is not None:
+        result.uuid = rec["uuid"]
+    else:
+        result.uuid = str(uuid.uuid4())
+    if "meta" in rec:
+        result.meta.update(context.object(rec["meta"]))
+
+    result._extended_component_id = context.object(rec["_extended_component_id"])
+
+    def fix_special_component_ids(ext_data):
```

Comment:
@astrofrog -- this is a hacky function I would like to eliminate, but I'm not sure how to rewrite the save/load logic to deal with the underlying problem. The basic problem is that when I restore a session, the ComponentIDs stored inside of an ExtendedComponent are no longer the same as the ComponentIDs for the actual components they reference. I think they are being created separately. This hacky function works, but would fail if there are multiple attributes with the same name/label.

Reply:
Yeah, that should work. Thanks for the idea!

Reply:
This solution involves saving ComponentID uuids into the session file and restoring them (if present) from the session file. I don't *think* this introduces and backwards-incompatibility problems (all tests pass) but does that seem okay to you?